### ISSUE TITLE
Memory leaks. from 80MB to 800MB for 11 days

### ISSUE THREAD

**Issue by halturin** [alacritty/alacritty#4806]
![image](https://user-images.githubusercontent.com/118860/108336845-c52d0700-71d4-11eb-8e5c-d7402e83b23c.png)

version: alacritty 0.7.1
uptime: 10:28:31  up 11 days 16:19,  1 user,  load average: 0.57, 0.25, 0.44

it only runs 'cava' during all this period.

if this is an intended behavior along with [Intensive CPU usage](https://github.com/alacritty/alacritty/issues/4528) just close this ticket :))

**Comment by chrisduerr** [comment]
How does this memory usage compare to a new terminal after running `yes` in it for a few seconds?

**Comment by halturin** [comment]
running 'yes' on a new terminal shows 173MB

**Comment by halturin** [comment]
btw. the terminal with 'cava' is still running and has 820MB
![image](https://user-images.githubusercontent.com/118860/108369818-46e45b00-71fc-11eb-8085-d8c965a08842.png)

**Comment by chrisduerr** [comment]
Are you resizing that terminal or moving it between monitors? Or is it just sitting there doing nothing?

**Comment by halturin** [comment]
terminal above started 11 days ago
![image](https://user-images.githubusercontent.com/118860/108370293-ca9e4780-71fc-11eb-8bf7-cee9fbbead85.png)
the second one - just a few seconds ago
as you may notice, they both have the same CPU consumption and 10 times differs in terms of memory usage.

![image](https://user-images.githubusercontent.com/118860/108370273-c40fd000-71fc-11eb-971b-68fb97d35c9c.png)

**Comment by chrisduerr** [comment]
Have you ever changed the config, especially the font size in the past 11 days?

**Comment by halturin** [comment]
no, I haven't even touched this window since I placed it in the corner.

**Comment by chrisduerr** [comment]
I'm talking about the configuration file of all Alacritty instances, not about changing anything about this window?

**Comment by halturin** [comment]
```
❯❯❯❯ ls -l  ./.config/nixpkgs/alacritty/alacritty.yml                                 
-rw-r--r-- 1 taras users 15625 Jan  7 17:02 ./.config/nixpkgs/alacritty/alacritty.yml
```

**Comment by chrisduerr** [comment]
How quickly does this increase for you? Is this something where you can notice a significant increase in a reasonably short period of time?

Also does it get worth the bigger the window and the smaller the font size? Being able to more easily reproduce this might help with just running things through valgrind and auto-detecting it.

**Comment by halturin** [comment]
all you need is to run 'cava' in the terminal and start playing something (music, youtube, whatever you want)... and just wait some time. this is the easiest way to reproduce this leak.

**Comment by chrisduerr** [comment]
Define 'some time'.

**Comment by halturin** [comment]
a year. just to be sure :rofl:

**Comment by halturin** [comment]
> Define 'some time'.

80MB for the last 9 hours.

![image](https://user-images.githubusercontent.com/118860/108405560-7f4a6000-7221-11eb-9f68-5c95a3218dbb.png)

so, its up to you how much time do you need for that.

**Comment by chrisduerr** [comment]
Not being able to reproduce this makes it somewhat difficult to troubleshoot. If there's a memory leak I'd expect a more noticable increase in memory size over that period of time.

**Comment by halturin** [comment]
Are you kidding me? 😄 Are you telling that isn't noticable? From 80 to 800 and still getting bigger it isn't noticable? 

Please, confirm it just to make me sure if I understand you correctly.

**Comment by chrisduerr** [comment]
I don't have the ability to run cava for a few hours at the moment. But I've been running an Alacritty instance for almost 9 hours and its memory usage is sitting at 90MiB.

**Comment by halturin** [comment]
Do you understand what cava does on the screen? If I were you I would test it by running something with the similar intensivity of output. Let it be 'yes', let it be whatever else what can produce such amount of output.

**Comment by nixpulvis** [comment]
> Do you understand what `cava` does on the screen? If I were you I would test it by running something with the similar intensivity of output. Let it be 'yes', let it be whatever else what can produce such amount of output.

Does `yes` have the same issue for you? Does the issue with `cava` go away when you quit `cava`?

**Comment by chrisduerr** [comment]
I am aware of what cava does, yes. But I cannot reproduce the issue.

Are you on a laptop and frequently closing its lid?

**Comment by halturin** [comment]
no, its a desktop pc. no suspend during this period.

**Comment by halturin** [comment]
>Does the issue with cava go away when you quit cava?

not sure if you understand how memory management works in the OS.

**Comment by halturin** [comment]
guys, is anyone of you can just start a new terminal and run there 'yes' command and wait an hour? it so simple, isn't it? I've just did it and got confirmation... leaking with the same speed - around 10MB/hour.

**Comment by chrisduerr** [comment]
Yes will naturally increase in memory usage at the beginning due to dynamically allocating the grid storage. After that it stops.

**Comment by halturin** [comment]
how long did you run it to make this conclusion? how much memory will be used before it stops leaking?

**Comment by chrisduerr** [comment]
That depends on your configuration.

**Comment by halturin** [comment]
feel free to try it with mine https://github.com/halturin/nixos-config/blob/master/.config/nixpkgs/alacritty/alacritty.yml (its almost default configuration)

**Comment by nixpulvis** [comment]
> not sure if you understand how memory management works in the OS.

I can assure you that I do understand. But that doesn't change the fact I'm curious how this behavior you're seeing might be impacted by different things, since I'm also unable to reproduce this running `yes` for many minutes and tracking it's terminal's memory.

**Comment by halturin** [comment]
@nixpulvis stopping 'cava' process can not reduce the memory consumption of the terminal. it only can stop the leaking. I thought it is an obvious thing.

**Comment by nixpulvis** [comment]
@halturin OK, how about you run `valgrind` and debug this more yourself please.

**Comment by frnco** [comment]
> a year. just to be sure 

I laughed at this. A lot. :rofl:

But about the leak, if it takes one year I'd guess it doesn't affect that many people, so I'd already consider this an edge-case from the start. It's only a matter of priorities, I mean, having `tmux` run smoothly is obviously much more relevant, as it's listed as the reason for not having tabs.

Anything that takes over a week of running continuously without any reboots or whatever is already on the realm of "could be a dependency bug, or a rust bug, or even a kernel bug..." Putting effort into something this uncertain seems a bit unreasonable. Now for running for an hour, that I believe it's a great idea, so I'm thinking I'll run a couple terms for a while and see how they fare, but I wouldn't consider anything over at most some 72 hours as relevant, and that's to be considerate to gentoo people because compiling. Other than gentoo, after 24 hours just restart the damn thing, alacritty runs in a damn window, meaning you need some DM and DE, and you probably also have a browser, so you most likely have a shitload of memory leaks everywhere already (Seriously have you used chrome the last 5 years? It's hellish).

**Comment by pbosab** [comment]
alacritty 0.9.0 (fed349aa)

have the same issue, but with ncmpcpp, and only when i switch to "music visualizer".
alacritty keep increasing cpu%, every 1 or 2 sec till 100% in under 1 min.

![2021-11-20-160152_4480x1080_scrot](https://user-images.githubusercontent.com/90603492/142731271-34a3f675-89f1-4048-932c-e934f7c82914.png)

**Comment by chrisduerr** [comment]
@pbosab How is that the same issue?

**Comment by kchibisov** [comment]
Just for the reference if anyone would ever wonder where the leak is, there's no leak, but it's a memory fragmentation error, so glibc can't release the buffers for whatever reason and it looks like it's leaking.

Alacritty use dynamic memory for its grid and resizes, etc could result in memory fragmentation which looks like a leak (it's simple to test due to multiwindow code, since closing it sometimes doesn't fully release the resources it had).
